File: backend/app/schemas/seller.py
import uuid
from datetime import date
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, field_validator


class SellerCreateRequest(BaseModel):
    name: str
    address: str
    phone: str | None = None
    gstin: str | None = None
    state: str | None = None
    state_code: str | None = None


class SellerUpdateRequest(BaseModel):
    name: str
    address: str
    phone: str | None = None
    gstin: str | None = None
    state: str | None = None
    state_code: str | None = None


class SellerResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    address: str
    phone: str | None
    gstin: str | None
    state: str | None
    state_code: str | None
    is_active: bool
    pending_balance: str = "0.00"


class OpeningBalanceRequest(BaseModel):
    request_id: uuid.UUID
    amount: Decimal
    occurred_on: date

    @field_validator("amount")
    @classmethod
    def validate_positive_amount(cls, value: Decimal) -> Decimal:
        if value <= 0:
            raise ValueError("amount must be greater than zero")
        return value


class PaymentRequest(BaseModel):
    request_id: uuid.UUID
    seller_id: uuid.UUID
    amount: Decimal
    occurred_on: date
    notes: str | None = None

    @field_validator("amount")
    @classmethod
    def validate_positive_amount(cls, value: Decimal) -> Decimal:
        if value <= 0:
            raise ValueError("amount must be greater than zero")
        return value


class BalanceAdjustmentRequest(BaseModel):
    request_id: uuid.UUID
    direction: Literal["INCREASE", "DECREASE"]
    amount: Decimal
    occurred_on: date
    notes: str | None = None

    @field_validator("amount")
    @classmethod
    def validate_positive_amount(cls, value: Decimal) -> Decimal:
        if value <= 0:
            raise ValueError("amount must be greater than zero")
        return value


class SellerTransactionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    entry_type: str
    amount: Decimal
    occurred_on: date
    notes: str | None


class SellerLedgerResponse(BaseModel):
    class InvoiceHistoryEntry(BaseModel):
        invoice_id: uuid.UUID
        invoice_number: str
        invoice_date: date
        grand_total: Decimal
        payment_mode: str
        status: str

    seller: SellerResponse
    transactions: list[SellerTransactionResponse]
    invoices: list[InvoiceHistoryEntry] = []

File: mobile/lib/models/seller.dart
class Seller {
  const Seller({
    required this.id,
    required this.name,
    required this.address,
    required this.phone,
    required this.gstin,
    required this.state,
    required this.stateCode,
    required this.isActive,
    required this.pendingBalance,
  });

  final String id;
  final String name;
  final String address;
  final String? phone;
  final String? gstin;
  final String? state;
  final String? stateCode;
  final bool isActive;
  final double pendingBalance;

  Seller copyWith({
    String? id,
    String? name,
    String? address,
    String? phone,
    String? gstin,
    String? state,
    String? stateCode,
    bool? isActive,
    double? pendingBalance,
  }) {
    return Seller(
      id: id ?? this.id,
      name: name ?? this.name,
      address: address ?? this.address,
      phone: phone ?? this.phone,
      gstin: gstin ?? this.gstin,
      state: state ?? this.state,
      stateCode: stateCode ?? this.stateCode,
      isActive: isActive ?? this.isActive,
      pendingBalance: pendingBalance ?? this.pendingBalance,
    );
  }

  factory Seller.fromJson(Map<String, dynamic> json) {
    return Seller(
      id: json['id'].toString(),
      name: json['name'] as String? ?? '',
      address: json['address'] as String? ?? '',
      phone: json['phone'] as String?,
      gstin: json['gstin'] as String?,
      state: json['state'] as String?,
      stateCode: json['state_code'] as String?,
      isActive: json['is_active'] as bool? ?? true,
      pendingBalance: _toDouble(json['pending_balance']),
    );
  }

  static double _toDouble(Object? value) {
    if (value is num) {
      return value.toDouble();
    }
    return double.tryParse(value?.toString() ?? '') ?? 0;
  }
}
